libpages/server.go

```diff
@@ -271,10 +272,61 @@ const (
 	gracefulShutdownTimeout = 16 * time.Second
 	httpReadHeaderTimeout   = 8 * time.Second
 	httpIdleTimeout         = 1 * time.Minute
-	stagingDiskCacheName    = "./kbp-cert-cache-dev"
+	stagingDiskCacheName    = "./kbp-cert-cache-staging"
 	prodDiskCacheName       = "./kbp-cert-cache"
 )
 
+func (s *Server) redirectHandlerFunc(w http.ResponseWriter, req *http.Request) {
+	// URL.RequestURI(), unlike the Request-URI field in HTTP header, does not
+	// include protocol and host. Instead, it's always constructed in the
+	// `http` package as `path?query` for HTTP(s) URLs (when Opaque field is
+	// empty).
+	if len(req.URL.Opaque) > 0 {
+		// This shouldn't happen for HTTP(s) URLs.
+		s.config.Logger.Warn("len(URL.Opaque)>0",
+			zap.String("url", req.URL.String()),
+			zap.String("Opaque", req.URL.Opaque))
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	http.Redirect(w, req, "https://"+req.Host+req.URL.RequestURI(),
+		// Using 307 (Temporary Redirect) here, 1) instead of 302 to require
+		// same method to be used; and 2) instead of 308/301 to avoid pollute
+		// visitors' browser cache in case kbp user wants to use the domain in
+		// a way that needs HTTP not to be redirected to HTTPS.
+		http.StatusTemporaryRedirect)
```

Comment:
https://support.google.com/webmasters/answer/6073543?hl=en

Standard is to use 301 redirects.

Reply:
Thanks for the link! First that's not a standard, but a best-practice guide. Second, this is different in the sense that we are not redirecting our own website; we are redirecting user's domains. 301 saves one round-trip, but when the cache pollution is a problem for the domain owner, there's pretty much nothing you can do.

I think it might make sense in the future to switch to 308 (just like 301) if opted-in through `.kbp_config`. But that's slightly more work. WDYT? If that sounds good to you, I'll create a ticket.

Reply:
Sounds good; thanks! FWIW it'd only be an extra roundtrip for every first visit. Subsequent requests in the same tab would go through HTTPs as long as no `http://` absolute address is in the HTML.